src/node/waveshaper.rs

```diff
@@ -0,0 +1,593 @@
+        );
+
+        let upsampler_x4 = FftFixedInOut::<f32>::new(
+            sample_rate as usize,
+            sample_rate as usize * 4,
+            512,
+            channels_x4,
+        );
+
+        let downsampler_x4 = FftFixedInOut::<f32>::new(
+            sample_rate as usize,
+            sample_rate as usize / 4,
+            128,
+            channels_x4,
+        );
+
+        Self {
+            sample_rate,
+            oversample,
+            curve,
+            curve_set,
+            upsampler_x2,
+            upsampler_x4,
+            downsampler_x2,
+            downsampler_x4,
+            channels_x2,
+            channels_x4,
+            receiver,
+        }
+    }
+
+    #[inline]
+    fn no_process(&self, input: &AudioBuffer, output: &mut AudioBuffer) {
+        for (i_data, o_data) in input.channels().iter().zip(output.channels_mut()) {
+            for (&i, o) in i_data.iter().zip(o_data.iter_mut()) {
+                *o = i;
+            }
+        }
+    }
+
+    #[inline]
+    fn process_none(&self, input: &AudioBuffer, output: &mut AudioBuffer) {
+        for (i_data, o_data) in input.channels().iter().zip(output.channels_mut()) {
+            o_data.copy_from_slice(&i_data[..]);
+        }
+    }
+
+    #[inline]
+    fn process_2x(&mut self, input: &AudioBuffer, output: &mut AudioBuffer) {
+        let wave_in = input.channels();
+
+        let up_wave_in = self.upsampler_x2.process(wave_in).unwrap();
+        let mut up_wave_out = up_wave_in.clone();
+
+        for (i_data, o_data) in up_wave_in.iter().zip(&mut up_wave_out) {
+            for (&i, o) in i_data.iter().zip(o_data.iter_mut()) {
+                *o = self.tick(i);
+            }
+        }
+
+        let wave_out = self.downsampler_x2.process(&up_wave_out).unwrap();
+
+        for (i_data, o_data) in wave_out.iter().zip(output.channels_mut()) {
+            for (&i, o) in i_data.iter().zip(o_data.iter_mut()) {
+                *o = i;
+            }
+        }
+    }
+
+    #[inline]
+    fn process_4x(&mut self, input: &AudioBuffer, output: &mut AudioBuffer) {
+        let wave_in = input.channels();
+
+        let up_wave_in = self.upsampler_x4.process(wave_in).unwrap();
+        let mut up_wave_out = up_wave_in.clone();
+
+        for (i_data, o_data) in up_wave_in.iter().zip(&mut up_wave_out) {
+            for (&i, o) in i_data.iter().zip(o_data.iter_mut()) {
+                *o = self.tick(i);
+            }
+        }
+
+        let wave_out = self.downsampler_x4.process(&up_wave_out).unwrap();
+
+        for (i_data, o_data) in wave_out.iter().zip(output.channels_mut()) {
+            for (&i, o) in i_data.iter().zip(o_data.iter_mut()) {
+                *o = i;
+            }
+        }
+    }
+
+    #[inline]
+    fn tick(&self, input: f32) -> f32 {
+        let n = self.curve.len() as f32;
+        if n == 0. {
+            return 0.;
+        }
+        let v = (n - 1.) / 2.0 * (input + 1.);
+
+        match v {
```

Comment:
This match statement is rather odd, you are matching `v` to itself. Better to write it as a set of if-statements.